pkg/cache/external/scheduler_cache.go

```diff
@@ -217,30 +217,21 @@ func (cache *SchedulerCache) AddPod(pod *v1.Pod) error {
 	defer cache.lock.Unlock()
 
 	currState, ok := cache.podsMap[key]
-	switch {
-	case ok && cache.isAssumedPod(key):
-		if currState.Spec.NodeName != pod.Spec.NodeName {
-			// The pod was added to a different node than it was assumed to.
-			log.Logger().Warn("inconsistent pod location",
-				zap.String("assumedLocation", pod.Spec.NodeName),
-				zap.String("actualLocation", currState.Spec.NodeName))
-
-			// Clean this up.
-			err = cache.removePod(currState)
-			if err != nil {
-				log.Logger().Debug("node not in cache",
-					zap.Error(err))
+	if ok {
+		// pod exists
+		if cache.isAssumedPod(key) {
```

Comment:
> in the default scheduler side, it only updates the pod when the pod.sepc.NodeName changes and the pod is assumed. In our code, we do the update no matter if the pod is assumed or not.

Original code then gets out of sync whenever a pod is assumed. There is no reason to do this, and in the case where the pod gets I assumed we are now stale. We should keep the new code. 

> Another thing, when the pod node name is unchanged, and the pod is assumed, it deletes the pod from the cache. [cache.go#489](https://github.com/kubernetes/kubernetes/blob/3866cb91f22da6eb49dab10dd5c33385690b57b4/pkg/scheduler/internal/cache/cache.go#L489). We no longer have this logic anywhere.

This is even worse. Now we lose a pod entirely while we have an active allocation. This also has run-on effects in plug-in mode because we don't explicitly control the bind phase and are reliant on informer updates to keep us synchronized. Again, new code is necessary. 

We are not a carbon-copy of upstream K8s and I'm not willing to keep obviously broken code simply because that's how upstream does it. Caches are meant to reflect the things they cache. Any conditional updates de-synchronize it and make it useless. The original code tries to be entirely too clever by half and resulted in a massive memory leak due to that complexity.